p2panda-net/src/sync/manager.rs

```diff
@@ -332,26 +332,23 @@ where
             session.remove(&sync_attempt.peer);
         }
 
-        if let Some(err) = err.downcast_ref() {
-            match err {
-                // If the sync attempt failed because of a connection error we want to retry up to
-                // `max_retry_attempts`. If error occurs after this we simply stop trying without
-                // informing the engine as it never knew the attempts were occurring.
-                SyncAttemptError::Connection => {
-                    warn!("sync attempt failed due to connection error");
-                    if sync_attempt.attempts <= self.config.max_retry_attempts {
-                        self.reschedule_attempt(sync_attempt).await?;
-                        return Ok(());
-                    }
-                }
-                SyncAttemptError::Sync(_) => {
-                    self.engine_actor_tx
-                        .send(ToEngineActor::SyncFailed {
-                            topic: Some(sync_attempt.topic),
-                            peer: sync_attempt.peer,
-                        })
-                        .await?;
-                }
+        if let Some(err) = err.downcast_ref::<SyncAttemptError>() {
+            // If the sync attempt failed for any reason we want to retry up to
+            // `max_retry_attempts`.
+            warn!("sync attempt failed: {err}");
+
+            // We need to inform the engine of the failed attempt so that the gossip buffer counter
+            // can be decremented.
+            self.engine_actor_tx
+                .send(ToEngineActor::SyncFailed {
+                    topic: Some(sync_attempt.topic.clone()),
+                    peer: sync_attempt.peer,
+                })
+                .await?;
+
+            if sync_attempt.attempts <= self.config.max_retry_attempts {
+                self.reschedule_attempt(sync_attempt).await?;
```

Comment:
@mycognosist do you mind clarifying for me what the interval for sync re-attempts are? I couldn't quite grok it looking at the configurable sync polling / resync interval values.

Reply:
Good question. The current logic pushes the failed attempt back into the regular sync queue. This means that the next sync attempt will occur as soon as it's pulled from the queue (if it's the only attempt in the queue it will be pulled immediately). In other words, there is currently no enforced interval. We only enforce a default interval of > 60 seconds for a _resync_ attempt (where the previous session was successful).

In answering your question, I see that this situation is sub-optimal. We'd really rather not be firing off five attempts in rapid succession after the initial failure.

Reply:
Yeh, that's what I was thinking about. The question came up when talking to Konstantin about the behaviour he was observing. It wasn't clear if sync sessions were being re-attempted on fail, but so quickly that they all failed and the max limit was reached before whatever caused the failure had a chance to recover.

Reply:
Would the polling interval already rate limit re-attempts to max one per second?

Reply:
The polling interval (of one second) only applies to the resync queue. The main queue is pulled from whenever an item is available.

Let me work on a solution.

Reply:
@sandreae I've opened a second PR with an overall refactor of the sync manager. It also introduces logic for the retry queue (with configurable intervals between attempts). 

My suggestion would be to merge this PR as-is and then review and merge [PR#714](https://github.com/p2panda/p2panda/pull/714).